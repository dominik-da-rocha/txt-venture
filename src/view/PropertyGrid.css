.PropertyGrid {
  display: flex;
  flex-direction: column;
  overflow-y: auto;
  height: 40rem;
  width: 30rem;
}

.PropertyGrid > .Object .Object,
.PropertyGrid > .Object .Value {
  margin-left: 1.5rem;
}

.PropertyGrid .Object {
  display: flex;
  flex-direction: column;
}

.PropertyGrid .Value {
  display: flex;
  flex-direction: row;
  align-items: baseline;
  align-content: stretch;
  background: whitesmoke;
}

.PropertyGrid .Object > .Expander {
  display: none;
}

.PropertyGrid .Object > .Expander ~ .ExpanderContent {
  visibility: collapse;
  height: 0;
  padding: 0;
  transition: height 0.5s ease;
}

.PropertyGrid .Object > .Expander:checked ~ .ExpanderContent {
  visibility: visible;
  height: auto;
  transition: height 0.5s ease;
}

.PropertyGrid .Object > .Expander ~ .ExpanderLabel::before {
  display: inline-block;
  width: 1rem;
  content: "▶";
}

.PropertyGrid .Object > .Expander:checked ~ .ExpanderLabel::before {
  content: "▼";
}

.PropertyGrid .Object > .ExpanderLabel .PropertyGrid .Value {
  flex-direction: row;
}

.PropertyGrid .Object > .ExpanderLabel,
.PropertyGrid .Object > .ExpanderLabel::before {
  display: inline-block;
  user-select: none;
  background: whitesmoke;
}

.PropertyGrid .Object > .ExpanderLabel::before {
  width: 1rem;
  padding-left: 0.25rem;
  padding-right: 0.25rem;
  text-align: center;
}

.PropertyGrid .Object > .ExpanderLabel:hover,
.PropertyGrid .Object > .ExpanderLabel:hover::before,
.PropertyGrid .Value:hover {
  background: #e0e0e0;
}

.PropertyGrid .Object > .ExpanderLabel:active,
.PropertyGrid .Object > .ExpanderLabel:active::before,
.PropertyGrid .Value:active {
  background: cornflowerblue;
  color: white;
}

.PropertyGrid .Value > .ExpanderLabel {
  width: 30%;
  user-select: none;
}

.PropertyGrid .Value > .Editor {
  flex: 1;
  margin: 0.1rem;
}
