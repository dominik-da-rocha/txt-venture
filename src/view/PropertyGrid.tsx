import "./PropertyGrid.css";
import React, { ChangeEvent } from "react";
import Component, { ComponentProps, ComponentState } from "./Component";

export interface PropertyGridProps extends ComponentProps {
  object: any;
  onChange(object: any): void;
}

export interface PropertyGridState extends ComponentState {
  object: any;
  items: any;
}

export class PropertyGrid extends Component<
  PropertyGridProps,
  PropertyGridState
> {
  key: number = 0;

  constructor(props: PropertyGridProps) {
    super(props, {
      object: props.object,
      items: {},
    });
  }

  render() {
    this.key = 0;
    return (
      <div id={this.props.id} className={this.className("PropertyGrid")}>
        {this.renderRows(this.state.object, "object")}
      </div>
    );
  }

  renderRows(object: any, path: string): React.ReactNode {
    let items: any[] = [];
    for (let name in object) {
      if (!name.endsWith("_")) {
        let fullName = path + "." + name;
        let type = typeof object[name] === "object" ? "Object" : "Value";
        let key = "item-" + ++this.key;
        let id = this.makeId(key);
        items.push(
          <div key={key} className={type + " ExpanderContent"}>
            {this.renderExpander(type, id, key)}
            <label className="ExpanderLabel" htmlFor={id}>
              {name}
            </label>
            {this.renderValue(object[name], id, fullName)}
          </div>
        );
      }
    }
    return items;
  }

  renderExpander(type: string, id: string, key: string) {
    return type === "Object" ? (
      <input
        className="Expander"
        type="checkbox"
        id={id}
        defaultChecked={this.state.items[key] ? false : true}
        onChange={(event: ChangeEvent<HTMLInputElement>) => {
          let value = !event.target.checked;
          this.setState((state: PropertyGridState) => {
            state.items[key] = value;
            return state;
          });
        }}
      />
    ) : undefined;
  }

  renderValue(object: any, id: string, fullName: string): React.ReactNode {
    switch (typeof object) {
      case "boolean":
        return (
          <input
            id={id}
            className="Editor"
            type="checkbox"
            defaultChecked={object}
            onChange={(event: ChangeEvent<HTMLInputElement>) => {
              this.setObject(fullName, event.target.checked);
            }}
          />
        );
      case "string":
        return (
          <input
            id={id}
            className="Editor"
            type="text"
            defaultValue={object}
            onChange={(event: ChangeEvent<HTMLInputElement>) => {
              this.setObject(fullName, event.target.value);
            }}
          />
        );
      case "number":
        return (
          <input
            id={id}
            className="Editor"
            type="number"
            defaultValue={object.toString()}
            onChange={(event: ChangeEvent<HTMLInputElement>) => {
              this.setObject(fullName, event.target.valueAsNumber);
            }}
          />
        );
      case "object":
        return this.renderRows(object, fullName);
    }
  }
  
  setObject(fullName: string, value: string | boolean | number) {
    this.setState((state: PropertyGridState) => {
      let object: any = state;
      let path = fullName.split(".");
      path.forEach((name: string) => {
        if (name === path[path.length - 1]) {
          object[name] = value;
        } else {
          object = object[name];
        }
      });
      this.props.onChange(state.object);
      return state;
    });
  }
}

export default PropertyGrid;