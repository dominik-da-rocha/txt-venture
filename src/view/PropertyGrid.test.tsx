import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Renderer from "react-test-renderer";
import React from "react";
import PropertyGrid from "./PropertyGrid";

Enzyme.configure({ adapter: new Adapter() });

class TestChild {
  value: boolean = false;
}

class TestObject {
  state: boolean = false;
  text: string = "test";
  count: number = 1;
  private_: string = "this is private";
  child: TestChild = new TestChild();
}

describe("PropertyGrid", () => {
  let test = new TestObject();
  it("renders without crashing", () => {
    expect(() => {
      shallow(
        <PropertyGrid
          id="test"
          object={test}
          onChange={(test: TestObject) => {}}
        />
      );
    }).not.toThrow();
  });

  it("renders snapshot", () => {
    let test = new TestObject();
    const tree = Renderer.create(
      <PropertyGrid
        id="test"
        className="Test"
        key="test"
        object={test}
        onChange={(test: TestObject) => {}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("changes expander", () => {
    let test = new TestObject();
    let wrapper = shallow(
      <PropertyGrid
        id="test"
        object={test}
        onChange={(test: TestObject) => {}}
      />
    );
    expect((wrapper.state() as any).items).toBeDefined();
    expect((wrapper.state() as any).items["item-4"]).toBeUndefined();

    wrapper
      .find("#test-item-4")
      .simulate("change", { target: { checked: false } });
    expect((wrapper.state() as any).items["item-4"]).toBeDefined();
    expect((wrapper.state() as any).items["item-4"]).toBeTruthy();

    wrapper
      .find("#test-item-4")
      .simulate("change", { target: { checked: true } });
    expect((wrapper.state() as any).items["item-4"]).toBeFalsy();
  });

  it("changes boolean value", () => {
    let test = new TestObject();
    let wrapper = shallow(
      <PropertyGrid
        id="test"
        object={test}
        onChange={(test: TestObject) => {}}
      />
    );
    expect((wrapper.state() as any).object.child.value).toBeFalsy();

    wrapper
      .find("#test-item-5")
      .simulate("change", { target: { checked: true } });
    expect((wrapper.state() as any).object.child.value).toBeTruthy();
  });

  it("changes string value", () => {
    let test = new TestObject();
    let wrapper = shallow(
      <PropertyGrid
        id="test"
        object={test}
        onChange={(test: TestObject) => {}}
      />
    );
    expect((wrapper.state() as any).object.text).toBe("test");

    wrapper
      .find("#test-item-2")
      .simulate("change", { target: { value: "Hello World" } });
    expect((wrapper.state() as any).object.text).toBe("Hello World");
  });

  it("changes number value", () => {
    let test = new TestObject();
    let wrapper = shallow(
      <PropertyGrid
        id="test"
        object={test}
        onChange={(test: TestObject) => {}}
      />
    );
    expect((wrapper.state() as any).object.count).toBe(1);

    wrapper
      .find("#test-item-3")
      .simulate("change", { target: { valueAsNumber: 123 } });
    expect((wrapper.state() as any).object.count).toBe(123);
  });
});
