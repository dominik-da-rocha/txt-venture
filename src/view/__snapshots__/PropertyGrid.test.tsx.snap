// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyGrid renders snapshot 1`] = `
<div
  className="Component PropertyGrid Test"
  id="test"
>
  <div
    className="Value ExpanderContent"
  >
    <label
      className="ExpanderLabel"
      htmlFor="test-item-1"
    >
      state
    </label>
    <input
      className="Editor"
      defaultChecked={false}
      id="test-item-1"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="Value ExpanderContent"
  >
    <label
      className="ExpanderLabel"
      htmlFor="test-item-2"
    >
      text
    </label>
    <input
      className="Editor"
      defaultValue="test"
      id="test-item-2"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="Value ExpanderContent"
  >
    <label
      className="ExpanderLabel"
      htmlFor="test-item-3"
    >
      count
    </label>
    <input
      className="Editor"
      defaultValue="1"
      id="test-item-3"
      onChange={[Function]}
      type="number"
    />
  </div>
  <div
    className="Object ExpanderContent"
  >
    <input
      className="Expander"
      defaultChecked={true}
      id="test-item-4"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="ExpanderLabel"
      htmlFor="test-item-4"
    >
      child
    </label>
    <div
      className="Value ExpanderContent"
    >
      <label
        className="ExpanderLabel"
        htmlFor="test-item-5"
      >
        value
      </label>
      <input
        className="Editor"
        defaultChecked={false}
        id="test-item-5"
        onChange={[Function]}
        type="checkbox"
      />
    </div>
  </div>
</div>
`;
